lib/koala/api/graph_batch_api.rb

```diff
@@ -41,67 +41,106 @@ def graph_call_in_batch(path, args = {}, verb = "get", options = {}, &post_proce
-        return [] unless batch_calls.length > 0
+        return [] if batch_calls.empty?
+
         # Turn the call args collected into what facebook expects
-        args = {}
-        args["batch"] = JSON.dump(batch_calls.map { |batch_op|
-          args.merge!(batch_op.files) if batch_op.files
-          batch_op.to_batch_params(access_token, app_secret)
-        })
+        args = { 'batch' => batch_args }
+        batch_calls.each do |call|
+          args.merge! call.files || {}
+        end
 
-        batch_result = graph_call_outside_batch('/', args, 'post', http_options) do |response|
-          unless response
-            # Facebook sometimes reportedly returns an empty body at times
-            # see https://github.com/arsduo/koala/issues/184
-            raise BadFacebookResponse.new(200, '', "Facebook returned an empty body")
-          end
+        graph_call_outside_batch('/', args, 'post', http_options, &handle_response)
+      end
+
+      def handle_response
+        lambda do |response|
+          raise bad_response if response.nil?
+          response.map(&generate_results)
+        end
+      end
 
-          # map the results with post-processing included
-          index = 0 # keep compat with ruby 1.8 - no with_index for map
-          response.map do |call_result|
-            # Get the options hash
-            batch_op = batch_calls[index]
-            index += 1
-
-            raw_result = nil
-            if call_result
-              parsed_headers = if call_result.has_key?('headers')
-                call_result['headers'].inject({}) { |headers, h| headers[h['name']] = h['value']; headers}
-              else
-                {}
-              end
-
-              if (error = check_response(call_result['code'], call_result['body'].to_s, parsed_headers))
-                raw_result = error
-              else
-                # (see note in regular api method about JSON parsing)
-                body = JSON.load("[#{call_result['body'].to_s}]")[0]
-
-                # Get the HTTP component they want
-                raw_result = case batch_op.http_options[:http_component]
-                when :status
-                  call_result["code"].to_i
-                when :headers
-                  # facebook returns the headers as an array of k/v pairs, but we want a regular hash
-                  parsed_headers
-                else
-                  body
-                end
-              end
-            end
-
-            # turn any results that are pageable into GraphCollections
-            # and pass to post-processing callback if given
-            result = GraphCollection.evaluate(raw_result, @original_api)
-            if batch_op.post_processing
-              batch_op.post_processing.call(result)
-            else
-              result
-            end
+      def generate_results
+        index = 0
+        lambda do |call_result|
+          batch_op     = batch_calls[index]; index += 1
```

Comment:
`each_with_index` doesn't go back to the version of Ruby being supported.  Yes, two lines are abnormal, and I wouldn't normally do it.  In this case, it just seemed to communicate better that two two segments were intertwined (because `each_with_index` wasn't available).  I find `index` or iterators are tough to track; meaning, it's hard to assert where/what it applies to.